app/handlers/profiles/settings/username.go

```diff
@@ -25,56 +25,67 @@ func NameChangeSubmitHandler(c *gin.Context) {
-		// check if username already taken
-		if services.DB.QueryRow("SELECT 1 FROM users WHERE username_safe = ?", uu.SafeUsername(username)).
-			Scan(new(int)) != sql.ErrNoRows {
-			sessions.AddMessage(c, msg.ErrorMessage{lu.T(c, "Username taken.")})
-			sessions.GetSession(c).Save()
-
-			c.Redirect(302, "/u/"+strconv.Itoa(int(sessions.GetContext(c).User.ID)))
+	isDonor := (ctx.User.Privileges & common.UserPrivilegeDonor) != 0
+	if !isDonor {
+		if !hasFreeUsernameChange {
+			tu.Resp403(c)
 			return
 		}
+		// if they are not a donor, consume their free username change
+		hasFreeUsernameChange = false
+	}
 
-		err := uu.ValidateUsername(username)
-		if err != "" {
-			sessions.AddMessage(c, msg.ErrorMessage{lu.T(c, err)})
-			sessions.GetSession(c).Save()
-
-			c.Redirect(302, "/u/"+strconv.Itoa(int(sessions.GetContext(c).User.ID)))
-			return
-		}
-
-		// update username
-		services.DB.Exec("UPDATE users SET username = ?, username_safe = ? WHERE id = ?", username, uu.SafeUsername(username), sessions.GetContext(c).User.ID)
-
-		logErr := uu.AddToUserNotes(fmt.Sprintf("Username change (self): %s -> %s", sessions.GetContext(c).User.Username, username), sessions.GetContext(c).User.ID)
-		if logErr != nil {
-			sessions.AddMessage(c, msg.ErrorMessage{lu.T(c, "Something went wrong.")})
-			sessions.GetSession(c).Save()
-
-			slog.Error("Error adding to user notes", "error", logErr.Error())
-
-			c.Redirect(302, "/u/"+strconv.Itoa(int(sessions.GetContext(c).User.ID)))
-			return
-		}
+	if c.PostForm("name") == "" {
+		sessions.AddMessage(c, msg.ErrorMessage{lu.T(c, "You have supplied an empty username.")})
+		sessions.GetSession(c).Save()
+		c.Redirect(302, "/u/"+strconv.Itoa(int(ctx.User.ID)))
+		return
+	}
 
-		services.RD.Publish("api:change_username", strconv.Itoa(int(sessions.GetContext(c).User.ID)))
+	username := strings.TrimSpace(c.PostForm("name"))
+	err := uu.ValidateUsername(username)
+	if err != "" {
+		sessions.AddMessage(c, msg.ErrorMessage{lu.T(c, err)})
+		sessions.GetSession(c).Save()
+		c.Redirect(302, "/u/"+strconv.Itoa(int(ctx.User.ID)))
+		return
+	}
+	
+	safeUsername := uu.SafeUsername(username)
 
-		sessions.AddMessage(c, msg.SuccessMessage{lu.T(c, "Username changed")})
+	// check if username already taken
+	if services.DB.QueryRow("SELECT 1 FROM users WHERE username_safe = ?", safeUsername).Scan(new(int)) != sql.ErrNoRows {
+		sessions.AddMessage(c, msg.ErrorMessage{lu.T(c, "Username taken.")})
 		sessions.GetSession(c).Save()
+		c.Redirect(302, "/u/"+strconv.Itoa(int(ctx.User.ID)))
+		return
+	}
+
+	// update username
+	services.DB.Exec("UPDATE users SET username = ?, username_safe = ?, has_free_username_change = ? WHERE id = ?", username, safeUsername, hasFreeUsernameChange, ctx.User.ID)
 
-		c.Redirect(302, "/u/"+strconv.Itoa(int(sessions.GetContext(c).User.ID)))
-	} else {
+	logErr := uu.AddToUserNotes(fmt.Sprintf("Username change (self): %s -> %s", ctx.User.Username, username), ctx.User.ID)
```

Comment:
possible to add donor vs used free name change here in the log? maybe nice to have

Reply:
yeah could do